File: alg/matrix.go
func add (a *matrix) (b *matrix) (c *matrix) {
  for i, m3x := 0, 0; i < len(a.ele); i += a.stride {
    for m2r0 := 0; m2r0 < b.stride; m2r0++ {
      c.ele[m3x] = a.ele[m3x] + b.ele[m3x]
      m1x++
    }
  }
  return c
}

func multiply(a *matrix) (b *matrix) (c *matrix) {
  for i, m3x := 0, 0; i < len(a.ele); i += a.stride {
    for m2r0 := 0; m2r0 < b.stride; m2r0++ {
      for m1x, m2x := m1c0, m2r0; m2x < len(b.ele); m2x += b.stride {
          c.ele[m3x] += a.ele[m1x] * b.ele[m2x]
          m1x++
      }
      m3x++
    }
  }
  return c
}

func custom(a *matrix) (b *matrix) (c *matrix) {
  for i, m3x := 0, 0; i < len(a.ele); i += a.stride {
    for m2r0 := 0; m2r0 < b.stride; m2r0++ {
      c.ele[m3x] = (max(a.ele[m3x],b.ele[m3x]) +
                   pow(a.ele[m3x], 2) +
                   pow(b.ele[m3x],2))
      m1x++
    }
  }
  return c
}

File: alg/matrix.c
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#define MAX(a,b) (((a)>(b))?(a):(b))

void add(int** a, int** b, int** c, int s)
{
  for (int i = 0; i < s; i++)
    for (int j = 0; j < s; j++)
      c[i][j] = a[i][j] + b[i][j];
}

void multiply(int** a, int** b, int** c, int s)
{
  for(int i = 0; i < s; i++)
    for (int j = 0; j < s; j++)
      for (int k = 0; k < s; k++)
        c[i][j] += a[i][k] * b[k][j];
}

void custom(int** a, int** b, int** c, int s)
{
  for(int i = 0; i < s; i++)
    for (int j = 0; j < s; j++)
      c[i][j] = MAX(a[i][j], b[i][j]) +
                (int)pow(a[i][j], 2)  +
                (int)pow(b[i][j], 2);
}

int main(int argc, char** argv)
{
  int n = atoi(argv[1]);
  int **a = (int**) malloc(n*n*sizeof(int));
  int **b = (int**) malloc(n*n*sizeof(int));
  int **c = (int**) malloc(n*n*sizeof(int));

  for (int i = 0; i < n; i++) {
    a[i][i] = i*2;
    b[i][i] = i;
  }

  add(a, b, c, n);
  multiply(a, b, c, n);
  custom(a, b, c, n);

  printf("%d\n", c[0][0]);

  return 0;
}
